src/renamer.fs

```diff
@@ -316,40 +311,46 @@ module private RenamerImpl =
+    let assignTemporaryIds shaders =
         let numberOfUsedIdents = ref 0
-        let env1 = Env.Create([], false, newTemporaryId numberOfUsedIdents, fun env _ -> env)
+        let mutable env = Env.Create([], false, newTemporaryId numberOfUsedIdents, fun env _ -> env)
         for shader in shaders do
-            shader.code <- renameTopLevel shader.code env1
+            let newEnv, code = renList env renTopLevelName shader.code
+            env <- newEnv
+            shader.code <- code
+
+        for shader in shaders do
+            shader.code <- List.map (renTopLevelBody env) shader.code
+        !env.exportedNames
+
+    let rename shaders =
+        let mutable exportedNames = assignTemporaryIds shaders
 
         // Get data about the context
         let text = [for shader in shaders -> Printer.printText shader.code] |> String.concat "\0"
         let identTable = computeFrequencyIdentTable text
         let contextTable = computeContextTable text
 
-        let mutable exportedNames = !env1.exportedNames
+        // TODO: combine from all shaders
+        let forbiddenNames = (Seq.head shaders).forbiddenNames
+
+        let idents = identTable |> Array.toList
+                   |> List.filter (fun x -> not <| List.exists ((=) x) forbiddenNames)
+
+        let mutable env = Env.Create(idents, true, optimizeContext contextTable, shadowVariables)
+        env <- dontRenameList env options.noRenamingList
+        env.exportedNames := exportedNames
 
-        // Second rename: use the context.
-        // TODO: first go through all top-level declarations in all files. Then, visit the
-        // function bodies. In other words, split renameTopLevel in two.
+        // First, rename top-level values.
         for shader in shaders do
-            // TODO: combine forbiddenNames. If a name is forbidden in a file, it shouldn't be used in
-            // exported values of another file.
-            let idents = identTable |> Array.toList
-                       |> List.filter (fun x -> not <| List.exists ((=) x) shader.forbiddenNames)
-            // TODO: optimizeContext should give the same name to exported values of the same name
-            // (e.g. when two files declare the same uniform)
-            let env2 = Env.Create(idents, true, optimizeContext contextTable, shadowVariables)
-            let env2 = dontRenameList env2 options.noRenamingList
-
-            // TODO: merge the exportedNames, make sure they don't have duplicates.
-            env2.exportedNames := exportedNames
-            shader.code <- renameTopLevel shader.code env2
-            exportedNames <- !env2.exportedNames
-
-        exportedNames
+            let newEnv, code = renList env renTopLevelName shader.code
+            env <- newEnv
+            shader.code <- code
+
+        // Rename local variables.
+        for shader in shaders do
+            shader.code <- List.map (renTopLevelBody env) shader.code
+
+        !env.exportedNames
 
```

Comment:
I think you ended up replacing both uses of renameTopLevel by the same code?
It could be factored back into a function.

Reply:
Done